ffmpeg/sign_test.go

```diff
@@ -0,0 +1,175 @@
+	if !ok {
+		glog.Fatalf("Expecting media playlist in the input %s", in)
+	}
+
+	var wg sync.WaitGroup
+
+	concurrentSessions := 1
+	segCount := 0
+	realTimeSegCount := 0
+	srcDur := 0.0
+	var mu sync.Mutex
+	transcodeDur := 0.0
+	for i := 0; i < concurrentSessions; i++ {
+		wg.Add(1)
+		go func(k int, wg *sync.WaitGroup) {
+			tc := NewTranscoder()
+			for j, v := range pl.Segments {
+				if v == nil {
+					continue
+				}
+				u := path.Join(dir, v.URI)
+				in := &TranscodeOptionsIn{
+					Fname: u,
+					Accel: accel,
+				}
+				if Software != accel {
+					in.Device = devices[k%len(devices)]
+				}
+				profs2opts := func(profs []VideoProfile) []TranscodeOptions {
+					opts := []TranscodeOptions{}
+					for n, p := range profs {
+
+						oname := fmt.Sprintf("%s_%s_%d_%d_%d.ts", outPrefix, p.Name, n, k, j)
+						muxer := "mpegts"
+
+						o := TranscodeOptions{
+							Oname:        oname,
+							Profile:      p,
+							Accel:        accel,
+							AudioEncoder: ComponentOptions{Name: "copy"},
+							Muxer:        ComponentOptions{Name: muxer},
+						}
+						if bSign {
+							o.CalcSign = true
+						}
+						opts = append(opts, o)
+					}
+					return opts
+				}
+				out := profs2opts(profiles)
+				t := time.Now()
+				_, err := tc.Transcode(in, out)
+				end := time.Now()
+				if err != nil {
+					glog.Fatalf("Transcoding failed for session %d segment %d: %v", k, j, err)
+				}
+
+				fmt.Printf("%s,%d,%d,%0.4v,%0.4v\n", end.Format("2006-01-02 15:04:05.9999"), k, j, v.Duration, end.Sub(t).Seconds())
+
+				segTxDur := end.Sub(t).Seconds()
+				mu.Lock()
+				transcodeDur += segTxDur
+				srcDur += v.Duration
+				segCount++
+				if segTxDur <= v.Duration {
+					realTimeSegCount += 1
+				}
+				mu.Unlock()
+			}
+			tc.StopTranscoder()
+			wg.Done()
+		}(i, &wg)
+		time.Sleep(2300 * time.Millisecond) // wait for at least one segment before moving on to the next session
+	}
+	wg.Wait()
+	if segCount == 0 || srcDur == 0.0 {
+		glog.Fatal("Input manifest has no segments or total duration is 0s")
+	}
+	statsTable := tablewriter.NewWriter(os.Stderr)
+	signmode := "Sign Mode"
+	if bSign == false {
+		signmode = "No Sign Mode"
+	}
+	devmod := "Software"
+	if bNvidia {
+		devmod = "Nvidia"
+	}
+	stats := [][]string{
+		{"Sign Status", fmt.Sprintf("%v - %v", signmode, devmod)},
+		{"Total Segs Transcoded", fmt.Sprintf("%v", segCount)},
+		{"Real-Time Segs Transcoded", fmt.Sprintf("%v", realTimeSegCount)},
+		{"* Real-Time Segs Ratio *", fmt.Sprintf("%0.4v", float64(realTimeSegCount)/float64(segCount))},
+		{"Total Source Duration", fmt.Sprintf("%vs", srcDur)},
+		{"Total Transcoding Duration", fmt.Sprintf("%vs", transcodeDur)},
+		{"* Real-Time Duration Ratio *", fmt.Sprintf("%0.4v", transcodeDur/srcDur)},
+	}
+	statsTable.SetAlignment(tablewriter.ALIGN_LEFT)
+	statsTable.SetCenterSeparator("*")
+	statsTable.SetColumnSeparator("|")
+	statsTable.AppendBulk(stats)
+	statsTable.Render()
+}
+
+func TestSign_SWTranscoding(t *testing.T) {
```

Comment:
What are these unit-tests testing exactly?

Does it make more sense to have a testcase that generates signatures and compares them, similar to what is done in `signcomp_test.go`?

If the motive here was to not unit test, but have a module that can generate signatures for us - it makes sense to add an option in our transcoding tool inside `cmd/` to optionally generate signatures for us. WDYT?

Reply:
Also if we decide to keep this as a test-case, we'll need to add a `nvidia` tag in the file, so that our CI pipeline which doesn't have GPUs is able to skip running these testcases.

Reply:
Right, I removed benchmarking test code and added all test code(generate and compare) into test_sign.go file.